velox/tool/trace/PartitionedOutputReplayer.cpp

```diff
@@ -0,0 +1,172 @@
+  return std::move(future).value();
+}
+} // namespace
+
+void consumeAllData(
+    const std::shared_ptr<exec::OutputBufferManager>& bufferManager,
+    const std::string& taskId,
+    uint32_t numPartitions,
+    folly::Executor* executor,
+    std::function<void(uint32_t, std::unique_ptr<folly::IOBuf>)> consumer) {
+  std::vector<std::thread> consumerThreads;
+  std::vector<int64_t> sequences;
+  consumerThreads.reserve(numPartitions);
+  sequences.reserve(numPartitions);
+  sequences.resize(numPartitions);
+  for (uint32_t i = 0; i < numPartitions; i++) {
+    consumerThreads.push_back(std::thread([&, partition = i]() {
+      bool finished{false};
+      while (!finished) {
+        std::vector<std::unique_ptr<folly::IOBuf>> pages;
+        {
+          pages = getData(
+              bufferManager, taskId, partition, sequences[partition], executor);
+        }
+        for (auto& page : pages) {
+          if (page) {
+            consumer(partition, std::move(page));
+            sequences[partition]++;
+          } else {
+            // Null page indicates this buffer is finished.
+            bufferManager->deleteResults(taskId, partition);
+            finished = true;
+          }
+        }
+      }
+    }));
+  }
+
+  for (auto& thread : consumerThreads) {
+    thread.join();
+  }
+}
+
+PartitionedOutputReplayer::PartitionedOutputReplayer(
+    const std::string& rootDir,
+    const std::string& taskId,
+    const std::string& nodeId,
+    const int32_t pipelineId,
+    const std::string& operatorType,
+    std::function<void(uint32_t, std::unique_ptr<folly::IOBuf>)> consumer)
+    : OperatorReplayerBase(rootDir, taskId, nodeId, pipelineId, operatorType),
+      originalNode_(dynamic_cast<const core::PartitionedOutputNode*>(
+          core::PlanNode::findFirstNode(
+              planFragment_.get(),
+              [this](const core::PlanNode* node) {
+                return node->id() == nodeId_;
+              }))),
+      consumer_(consumer) {
+  VELOX_CHECK_NOT_NULL(originalNode_);
+}
+
+RowVectorPtr PartitionedOutputReplayer::run() {
+  auto task = Task::create(
+      "local://partitioned-output-replayer",
+      core::PlanFragment{createPlan()},
+      0,
+      createQueryContext(queryConfigs_, executor_.get()),
+      Task::ExecutionMode::kParallel);
+  task->start(maxDrivers_);
+
+  consumeAllData(
+      bufferManager_,
+      task->taskId(),
+      originalNode_->numPartitions(),
+      executor_.get(),
+      consumer_);
+  return nullptr;
+}
+
+core::PlanNodePtr PartitionedOutputReplayer::createPlanNode(
+    const core::PlanNode* /* unused */,
+    const core::PlanNodeId& /* unused */,
+    const core::PlanNodePtr& /* unused */) const {
+  VELOX_UNREACHABLE(
+      "createPlanNode should not be called from PartitionedOutputReplayer");
+}
+
+core::PlanNodePtr PartitionedOutputReplayer::createPlan() const {
+  return PlanBuilder()
+      .traceScan(nodeDir_, exec::trace::getDataType(planFragment_, nodeId_))
```

Comment:
Why we can't use default replayNodeFactory to handle this? Thanks!

Reply:
We need a consumer stage so custom will be cleaner.

Reply:
I mean why can't implement createPlanNode for partition output replay operator and use replayNodeFactory to create the node? What's the custom stage?